### PR TITLE
prevent role and set duplicates

### PR DESCRIPTION
**Pull request by fairclothjm**
There is an edge case where a user can add an LDAP user or service account to more than one role or set. This is only possible after the backend has been reloaded.

Roles and sets can be defined as hierarchical paths (i.e. a graph). So we update the initialization path to perform a breadth-first search on the plugin's storage for both roles and sets. If we find an entry, we add it to the managed users map.

We attempt to make as few reads from storage as possible by returning the static roles from the `loadManagedUsers` func and passing them to the `initQueue` function instead of performing another lookup of each node in the graph. We do not want superfluous storage access on plugin initialization which could degrade Vault startup times.

### CODE REVIEW

File: rotation.go [hunk 1]
```diff
@@ -585,50 +572,3 @@ func (b *backend) popFromRotationQueueByKey(name string) (*queue.Item, error) {
 	}
 	return nil, queue.ErrEmpty
 }
-
-// getAllStaticRoles returns a slice of all static roles from storage
-func (b *backend) getAllStaticRoles(ctx context.Context, s logical.Storage) ([]string, error) {
```
Comment (fairclothjm) [inline rotation.go:590]: This was refactored to the generic `walkStoragePath` to be used for either roles or sets

File: backend.go [hunk 2]
```diff
@@ -149,10 +151,57 @@ type backend struct {
...
+		entryExists, err := walker(key)
+		if err != nil {
+			return fmt.Errorf("unable to read entry %q: %w", key, err)
+		}
+
+		if !entryExists && strings.HasSuffix(key, "/") {
+			// this is a directory
+			subKeys, err := s.List(ctx, path+key)
+			if err != nil {
+				return fmt.Errorf("unable to list keys: %w", err)
```
Comment (tomcf-hcp) [inline backend.go:183]: Is there ever a case where "foo/bar" would fail, but the "foo/bar/baz" would succeed and both of those would be "correct"?

Comment (fairclothjm) [reply backend.go:183]: An error returned from `storage.List()` means we could not access storage or we are in some bad state with the keys and we are dead in the water.

We have tests to cover the cases that I think you are describing:
https://github.com/hashicorp/vault-plugin-secrets-openldap/blob/156b280cd38c4d77e61d34839dad42396f9384f1/rotation_test.go#L48-L55

**Review by vinay-gopalan — APPROVED** [review]
LGTM!

**Pull request merged**